File: frontend/src/services/seed.ts
/**
 * Tipos del dominio y datos de respaldo del frontend.
 *
 * Tras conectar el backend, los datos reales vienen de la API (services/api.ts).
 * Aquí solo quedan: los **tipos** compartidos, y unos pocos valores usados como
 * defaults antes de iniciar sesión (VITALS, MEDICINES) o como datos de UI que aún
 * no tienen endpoint (PATIENTS — lista de "Mis pacientes" del médico, demo).
 */

// ---------- tipos del dominio ----------
export type Role = 'paciente' | 'medico' | 'admin'

export interface Session {
  role: Role
  nombres: string
  apellidos: string
  dni: string
  extra?: string
}

export interface Vitals {
  sangre: string
  alergias: string[]
  enfermedades: string[]
  medicacion: string[]
  contacto: { nombre: string; relacion: string; telefono: string }
  /** Solo lo recibe el paciente dueño desde GET /vitals. */
  emergencyCode?: string
}

export interface Patient {
  dni: string
  nombre: string
  edad: number
  sexo: string
  sangre: string
  cond: string
  ultimo: string
  eventos: number
  iniciales: string
  color: string
}

export interface LabRow {
  p: string
  v: string
  u: string
  r: string
  e: 'normal' | 'alto' | 'bajo'
}

export type Estudio =
  | { t: 'lab'; n: string; lab: string; rows: LabRow[] }
  | { t: 'img'; n: string; modalidad: string; equipo: string; hallazgos: string }

export interface Signos {
  pa: string
  fc: string
  fr: string
  temp: string
  sat: string
  peso: string
  talla: string
  imc: string
}

export interface Medicamento {
  nombre: string
  dosis: string
  indicacion: string
}

export interface ClinicalEvent {
  id: string
  ts: string
  tipo: string
  titulo: string
  medico: string
  colegiatura: string
  especialidad: string
  estab: string
  sev: string
  motivo: string
  diagnostico: string
  examen: string
  signos: Signos
  medicamentos: Medicamento[]
  indicaciones: string
  estudios: Estudio[]
  nuevo?: boolean
}

export interface Farmacia {
  nombre: string
  precio: number
  stock: string
  dist: string
}

export interface Medicine {
  id: string
  nombre: string
  principio: string
  forma: string
  categoria: string
  receta: boolean
  farmacias: Farmacia[]
}

export type TokenStatus = 'activa' | 'expirada' | 'agotada' | 'revocada'

export interface Token {
  code: string
  createdAt: number
  expiresAt: number
  durationMin: number
  uses: number
  usesLeft: number
  status: TokenStatus
  medico: string
}

export type AuditRole = 'medico' | 'emergencia' | 'paciente'

export interface AuditEntry {
  ts: number
  actor: string
  rol: AuditRole
  accion: string
  ref: string
  ip: string
  disp: string
}

export interface SessionDevice {
  id: string
  disp: string
  ip: string
  lugar: string
  actual: boolean
  ultimo: string
}

export interface RecetaItem {
  nombre: string
  dosis: string
  duracion: string
}

export interface Receta {
  id: string
  ts: string
  paciente: string
  medico: string
  estado: string
  nuevo?: boolean
  items: RecetaItem[]
}

export interface PendingStudy {
  id: string
  ts: string
  tipo: string
  nombre: string
  indicacion: string
  medico: string
  estado: string
}

// ---------- defaults / datos de UI ----------
/** Default mostrado antes de cargar los vitales reales desde la API. */
export const VITALS: Vitals = {
  sangre: 'O+ (O positivo)',
  alergias: ['Penicilina', 'Mariscos'],
  enfermedades: ['Asma bronquial'],
  medicacion: ['Salbutamol inhalador'],
  contacto: { nombre: 'María Quispe Ramos', relacion: 'Madre', telefono: '+51 998 123 456' },
}

/** Lista de "Mis pacientes" del médico (demo; aún sin endpoint). */
export const PATIENTS: Patient[] = [
  { dni: '45872136', nombre: 'Juan Carlos Pérez Quispe', edad: 35, sexo: 'M', sangre: 'O+', cond: 'Asma bronquial', ultimo: 'Hace 2 horas', eventos: 5, iniciales: 'JP', color: 'linear-gradient(150deg,#0d7d74,#0a5c55)' },
  { dni: '70112233', nombre: 'Lucía Fernanda Rojas Medina', edad: 27, sexo: 'F', sangre: 'A+', cond: 'Sin antecedentes', ultimo: 'Hace 5 días', eventos: 2, iniciales: 'LR', color: '#7c3aed' },
  { dni: '08456712', nombre: 'Pedro Antonio Huamán Soto', edad: 49, sexo: 'M', sangre: 'B+', cond: 'Hipertensión arterial', ultimo: 'Hace 2 semanas', eventos: 8, iniciales: 'PH', color: '#1d4ed8' },
]

/** Default del catálogo antes de cargar /medicines desde la API. */
export const MEDICINES: Medicine[] = [
  { id: 'm1', nombre: 'Salbutamol', principio: 'Salbutamol 100 mcg', forma: 'Inhalador · 200 dosis', categoria: 'Broncodilatador', receta: true,
    farmacias: [{ nombre: 'Boticas BTL', precio: 16.9, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Inkafarma', precio: 18.5, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Boticas Perú', precio: 17.2, stock: 'Pocas unidades', dist: '0.9 km' }, { nombre: 'Farmacias Universal', precio: 19.5, stock: 'En stock', dist: '2.1 km' }, { nombre: 'Mifarma', precio: 21.9, stock: 'En stock', dist: '0.6 km' }] },
  { id: 'm2', nombre: 'Budesonida', principio: 'Budesonida 200 mcg', forma: 'Inhalador · 120 dosis', categoria: 'Corticoide inhalado', receta: true,
    farmacias: [{ nombre: 'Inkafarma', precio: 42.9, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 45.5, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 39.9, stock: 'Pocas unidades', dist: '1.2 km' }, { nombre: 'Farmacias Universal', precio: 47.0, stock: 'En stock', dist: '2.1 km' }] },
  { id: 'm3', nombre: 'Paracetamol', principio: 'Paracetamol 500 mg', forma: 'Caja · 100 tabletas', categoria: 'Analgésico / Antipirético', receta: false,
    farmacias: [{ nombre: 'Inkafarma', precio: 8.9, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 9.5, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 6.5, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Boticas Perú', precio: 7.2, stock: 'En stock', dist: '0.9 km' }, { nombre: 'Farmacias Universal', precio: 8.0, stock: 'Pocas unidades', dist: '2.1 km' }] },
  { id: 'm4', nombre: 'Amoxicilina', principio: 'Amoxicilina 500 mg', forma: 'Caja · 21 cápsulas', categoria: 'Antibiótico', receta: true,
    farmacias: [{ nombre: 'Inkafarma', precio: 14.9, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 13.9, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 12.5, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Boticas Perú', precio: 15.0, stock: 'Agotado', dist: '0.9 km' }] },
  { id: 'm5', nombre: 'Ibuprofeno', principio: 'Ibuprofeno 400 mg', forma: 'Caja · 100 tabletas', categoria: 'Antiinflamatorio (AINE)', receta: false,
    farmacias: [{ nombre: 'Inkafarma', precio: 11.9, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 12.5, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 9.9, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Farmacias Universal', precio: 10.5, stock: 'En stock', dist: '2.1 km' }] },
  { id: 'm6', nombre: 'Loratadina', principio: 'Loratadina 10 mg', forma: 'Caja · 10 tabletas', categoria: 'Antihistamínico', receta: false,
    farmacias: [{ nombre: 'Inkafarma', precio: 7.5, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 6.9, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 5.9, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Boticas Perú', precio: 6.5, stock: 'Pocas unidades', dist: '0.9 km' }] },
  { id: 'm7', nombre: 'Omeprazol', principio: 'Omeprazol 20 mg', forma: 'Caja · 14 cápsulas', categoria: 'Inhibidor de bomba de protones', receta: false,
    farmacias: [{ nombre: 'Inkafarma', precio: 10.9, stock: 'En stock', dist: '0.4 km' }, { nombre: 'Mifarma', precio: 11.5, stock: 'En stock', dist: '0.6 km' }, { nombre: 'Boticas BTL', precio: 8.9, stock: 'En stock', dist: '1.2 km' }, { nombre: 'Farmacias Universal', precio: 9.9, stock: 'En stock', dist: '2.1 km' }] },
]

File: backend/app/seed.py
import time
from datetime import timedelta

from sqlalchemy.orm import Session

from . import data, models
from .security import hash_password
from .util import now_utc


def seed_if_empty(db: Session) -> None:
    if db.query(models.User).first():
        return

    for u in data.USERS:
        db.add(models.User(
            dni=u["dni"], password_hash=hash_password(u["password"]), role=u["role"],
            nombres=u["nombres"], apellidos=u["apellidos"], extra=u["extra"],
        ))

    for dni, r in data.RENIEC.items():
        db.add(models.ReniecRecord(dni=dni, **r))

    v = data.VITALS
    db.add(models.Vitals(
        patient_dni=data.PATIENT_DNI, sangre=v["sangre"], alergias=v["alergias"],
        enfermedades=v["enfermedades"], medicacion=v["medicacion"], contacto=v["contacto"],
    ))

    base = time.time()
    for i, e in enumerate(data.HISTORY):
        db.add(models.ClinicalEvent(
            id=e["id"], patient_dni=data.PATIENT_DNI, ts=e["ts"], tipo=e["tipo"], titulo=e["titulo"],
            medico=e["medico"], colegiatura=e["colegiatura"], especialidad=e["especialidad"],
            estab=e["estab"], sev=e["sev"], motivo=e["motivo"], diagnostico=e["diagnostico"],
            examen=e["examen"], signos=e["signos"], medicamentos=e["medicamentos"],
            indicaciones=e["indicaciones"], estudios=e["estudios"], nuevo=False,
            created_ts=base - i,
        ))

    for i, r in enumerate(data.RECETAS):
        db.add(models.Receta(
            id=r["id"], patient_dni=data.PATIENT_DNI, ts=r["ts"], paciente=r["paciente"],
            medico=r["medico"], estado=r["estado"], items=r["items"], nuevo=False,
            created_ts=base - i,
        ))

    for sdv in data.SESSIONS:
        db.add(models.SessionDevice(
            id=sdv["id"], user_dni=data.PATIENT_DNI, disp=sdv["disp"], ip=sdv["ip"],
            lugar=sdv["lugar"], actual=sdv["actual"], ultimo=sdv["ultimo"],
        ))

    now = now_utc()
    db.add(models.Token(code="HMPQ-7K2D-9F4X", patient_dni=data.PATIENT_DNI,
                        created_at=now - timedelta(minutes=8), expires_at=now + timedelta(seconds=892),
                        duration_min=30, uses=3, uses_left=2, status="activa", medico="—"))
    db.add(models.Token(code="HMPQ-3B8M-2P5T", patient_dni=data.PATIENT_DNI,
                        created_at=now - timedelta(days=3), expires_at=now - timedelta(days=3) + timedelta(minutes=30),
                        duration_min=30, uses=1, uses_left=0, status="expirada", medico="Dra. Ana Flores"))

    audits = [
        (now - timedelta(hours=2), "Dra. Ana Flores", "medico", "Consultó historial mediante token", "HMPQ-3B8M-2P5T", "190.233.45.18", "Chrome · Windows 11"),
        (now - timedelta(days=3), "Personal de emergencia", "emergencia", "Acceso en modo emergencia (datos vitales)", "EMG-45872136", "191.98.12.4", "Tablet · Hospital Loayza"),
        (now - timedelta(days=6), "Juan C. Pérez", "paciente", "Generó token médico", "HMPQ-9X1K-7L2N", "190.233.45.12", "Chrome · Windows 11"),
        (now - timedelta(days=9), "Dr. Luis Tello", "medico", "Consultó historial mediante token", "HMPQ-5R4D-8C3V", "200.48.225.9", "Edge · Windows 10"),
    ]
    for ts, actor, rol, accion, ref, ip, disp in audits:
        db.add(models.AuditEntry(ts=ts, patient_dni=data.PATIENT_DNI, actor=actor, rol=rol,
                                 accion=accion, ref=ref, ip=ip, disp=disp))

    db.commit()
